### ISSUE TITLE
Deal with ADMIXTOOLS crashes properly

### ISSUE THREAD

**Issue by bodkan** [bodkan/admixr#52]
From a user recently:

> in some cases but I have some problems running D-test with large numbers of comparisons:
> 
> I keep getting the following error message quite often but I do not fully understand why this happens:
> 
```
Warning message:

ADMIXTOOLS does not allow duplicated population names when calculating a single f4 or D statistic.
 
Returning an empty dataframe...
```
> 
> I am quite sure that my dataset does not contain any duplicates and I have been able to run it successfully by splitting up the population list into smaller sub-lists but this based on empirical trials and I do not really know how many populations it can handle. I wonder if you know what is wrong with my calculations and do you have a possible solution for such cases.

In this particular example, this happened because ADMIXTOOLS was killed by the OS for running out of memory. This causes the `log` file to be truncated and admixr is confused by that, reporting nonsense as the above.

Example log file:
[qpDstat__644628461.log](https://github.com/bodkan/admixr/files/4324917/qpDstat__644628461.log)

**Comment by bodkan** [comment]
Closed by [this](https://github.com/bodkan/admixr/commit/37d992c14bff5c118abb4b7cb4f2f553aed038c1) commit.

**Comment by dora-barle** [comment]
Hello Martin.
I believe I have this problem. My files are read and I can generate the snp_info.
Run the command for the first time:
result <- d(W = pops, X = "Pop1", Y = "Pop2", Z = "Pop3", data = snps), but I had 6 pops in my .ind file.
Then I edited the .ind file for three populations.
When I tried to run it again, I got the same error.

Warning message:
ADMIXTOOLS does not allow duplicated population names when calculating a single f4 or D statistic.
Returning an empty dataframe...
![ind](https://user-images.githubusercontent.com/83590787/227327323-7c5abd0f-2764-460e-a65e-caeb5cb21b78.png)
Try looking for qpDstat.log but I can't find it on the system Ubuntu

Do you know how i can solve this problem? Or could it be another problem with my data?

Thank you very much

Try looking for qpDstat.log but I can't find it on the system.

**Comment by dora-barle** [comment]
Hi,
I am trying to run admixr on Linux by following the manual (https://www.bodkan.net/admixr/articles/tutorial.html). However, I am not getting the expected test results for D or f4. I have tried various ways of inputting the populations, including assuming 3 and 6 in the ind file, as well as leaving the ind file with the names of each individual and then grouping them in R. However, none of these methods have worked for me. I am unsure of what could be causing the issue. 

I have two types of errors or warning: 
1. When analyzing three populations it appears:
Warning message:
ADMIXTOOLS does not allow duplicated population names when calculating a single f4 or D statistic.
Returning an empty dataframe...

2. When analyzing six populations it appears:
Error: The output file we got from ADMIXTOOLS is truncated.
Please examine the full output above check for errors.

I don't know what I could be doing wrong.

My files have the format shown in the attached screen prints for snp, geno, and ind. I used a script (convertVCFtoEigenstrat.sh) to convert them to a compatible format.

I also attach the .log files

Thank you very much.

![Erro_3](https://user-images.githubusercontent.com/83590787/227742890-98035b46-be4a-452d-bd33-c11c3bbfc950.png)
![Erro6](https://user-images.githubusercontent.com/83590787/227742894-da072baf-50ce-
![i
![geno](https://user-images.githubusercontent.com/83590787/227742904-37772ba4-ed4c-4469-8c8d-8c13bc9be09b.png)
[qpDstat__745985646-6pop.log](https://github.com/bodkan/admixr/files/11070215/qpDstat__745985646-6pop.log)
[qpDstat__1704914233-3pop.log](https://github.com/bodkan/admixr/files/11070216/qpDstat__1704914233-3pop.log)

![snp](https://user-images.githubusercontent.com/83590787/227742901-1f0716a2-c9c4-49e9-9d59-48b6008094a1.png)
nd](https://user-images.githubusercontent.com/83590787/227742898-c1e7a6e4-db59-4d78-b35d-e7b4d676aa0a.png)
4d5a-9bf1-ab820905d94d.png)
![Erro6](https://user-images.githubusercontent.com/83590787/227743057-d4a652c6-6049-4d72-ba63-97582113908b.png)

**Comment by bodkan** [comment]
Hi, thank you for submitting such an exhaustive issue, including those screenshots and log files (those in particular). 

I do have to say I'm having a bit of a trouble assigning all the screenshots to the issues you're having -- I assume the first (black, terminal) screenshot belong to the [qpDstat__1704914233-3pop.log](https://github.com/bodkan/admixr/files/11070216/qpDstat__1704914233-3pop.log) you've posted and the second terminal screenshot belongs to this [qpDstat__745985646-6pop.log](https://github.com/bodkan/admixr/files/11070215/qpDstat__745985646-6pop.log)?

Taking a look at the log files gives a hint of what's going on:

1. [qpDstat__1704914233-3pop.log](https://github.com/bodkan/admixr/files/11070216/qpDstat__1704914233-3pop.log)

```
[... snip ...]
nrows, ncols: 24 978712
***warning: repeated population: Pop1 Pop1 :  Pop2 Pop3 
***warning: repeated population: Pop2 Pop1 :  Pop2 Pop3 
***warning: repeated population: Pop3 Pop1 :  Pop2 Pop3 
##end of qpDstat:       10.527 seconds cpu      720.355 Mbytes in use
```

This suggests that each of the population quartets that are generated for ADMIXTOOLS computation have a duplicated population -- in cases like this, ADMIXTOOLS does not return any result because such quartets don't produce any interesting ABBA/BABA pattern (i.e. in `D(Pop1, Pop1; Pop2, Pop3)` you never get either ABBA or BABA because Pop1 is twice in the first two positions which therefore cannot differ in their allele states, the other quartets have similar problems).

You can test this even using the example from the admixr tutorial, just try to compute something like this:

```
d(W = c("Yoruba", "Vindija", "Chimp"), X = "Yoruba", Y = "Vindija", Z = "Chimp", data = snps)
```

and you get the same warning -- there is simply nothing interested for ADMIXTOOLS to compute, and admixr therefore returns nothing (because no numerical result can come from that analysis), giving a warning.

2. [qpDstat__745985646-6pop.log](https://github.com/bodkan/admixr/files/11070215/qpDstat__745985646-6pop.log)

Are you sure you posted the correct output log? Because this log file at the link you posted contains different information from your last screenshot (the 2nd black terminal screenshot)? In any case, in this screenshot I first see this error 

<img width="791" alt="image" src="https://user-images.githubusercontent.com/16516593/227934103-37d929bc-f78f-4fbc-b311-e867e040ec2d.png">

Like the above, this error again comes from ADMIXTOOLS (not admixr) -- see it originating [here](https://github.com/DReichLab/AdmixTools/blob/6831501f9967270cbb4e7fa2e13d6a7cc41f2be5/src/qpDstat.c#L1178). If this is persistent, maybe there's some issue with your data? In this situation, I would advise putting together a minimum sufficient example (`par` file) using raw ADMIXTOOLS and asking for help [here](https://github.com/DReichLab/AdmixTools). Unfortunately, I can't dedicate a lot of bandwidth to debugging ADMIXTOOLS-specific problems :( If you get this error even if you run `qpDstat` on the `par` file on the command-line (without admixr), then this is an ADMIXTOOLS problem.

Do you know what I'm saying above? Have you worked with `qpDstat` on the command-line before?

The other error reported in your 2nd black screenshot, so this one

<img width="486" alt="image" src="https://user-images.githubusercontent.com/16516593/227935530-32c8f680-47c8-4b82-a513-298e984a03bc.png">

looks exactly the same like the one in your _first log file_ (repeated populations, as I explained under 1. above) making me think that you might have got the log files/screenshots slightly mismatched?

**Comment by bodkan** [comment]
Hold on, I'm looking at your logs again and see this:

<img width="393" alt="image" src="https://user-images.githubusercontent.com/16516593/227936842-4e138fb6-fcb1-42d4-879f-aa32a9b5012e.png">

Only 3 blocks for jackknife can't be possibly right. There's definitely something wrong with the data which is why ADMIXTOOLS refuses to compute this.

Looking at your snp file, it looks a bit off -- the 3rd column is [supposed to be](https://github.com/DReichLab/AdmixTools/blob/master/convertf/README#L50) the genetic distance but your numbers can't be right, right? You have 8.00000, 2.00000, etc.

> I used a script (convertVCFtoEigenstrat.sh) to convert them to a compatible format.

I'm not sure what this script is supposed to be? If a collaborator or someone like that wrote it for you, perhaps you should talk to them.

**Comment by bodkan** [comment]
Overall, I don't think any of the issues are actually related to admixr. My guess (from what I've managed to put together from the files and screenshots you've submitted) is that 1. you're possibly asking ADMIXTOOLS to compute D statistics using quartets which don't really make sense (not possible to compute ABBA/BABA patterns on duplicated populations), and 2. you're data is probably not in a valid EIGENSTRAT format [as required by ADMIXTOOLS](https://github.com/DReichLab/AdmixTools/blob/master/convertf/README#L65).

**Comment by dora-barle** [comment]
Hello. 
Thank you for your detailed answer. You interpreted the screenshots well. 
I was also suspicious of the EIGENSTRAT files, but it was the way I found to generate the EIGENSTRAT from the VCF file.
I also tried to run Admixtools directly, but I was unable to do so.
I'll try generating the EIGENSTRAT files using another method and then I attempt the commands again.
Thank you very much for your help.
